### PR TITLE
alternator: vector search and streams on the same table

### LINKED ISSUE #7239: Alternator Stream should still be usable (for 24 hours) after disabling
The DynamoDB documentation https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html says that:
> If you disable a stream, any shards that are open will be closed. The data in the stream will continue to be readable for 24 hours. 

However, in Alternator, we don't allow this.  A new test `test_streams.py`::`test_streams_closed_next` tries to use shard iterators which were obtained before the stream was disabled, and the test shows that this is possible on DynamoDB, but generates an error on Alternator:
```
An error occurred (ResourceNotFoundException) when calling the GetRecords operation: cb8ff030-f74a-11ea-b0cc-000000000000
```

As far as I understand, there is no reason why this should not be allowed in Alternator as well; Disabling streaming does not (as far as I know) remove the existing stream data, so it should be just as usual as it was before. `ListStreams` should list the old stream (at least for 24 hours) - just `DescribeStream` should say it is `DISABLED`.

### PR DESCRIPTION
**Pull request by nyh**
Previously, Alternator Streams was limited to vnodes only while Vector Search was limited to tablets only. This was fixed and now both features are supported on a table based on tablets (which today is the default for new tables).

But right now, trying to enable both streams (with KEYS_ONLY) and vector search on the same table doesn’t work.

*  If you enable streams (with KEYS_ONLY) and then vector search, enabling the vector search fails with InternalServerError, stating that “The CDC log must meet the minimal requirements of Vector Search. This means that the CDC's TTL must be at least 86400 seconds (24 hours), and the CDC's delta mode must be set to 'full' or postimage must be enabled.”
* If you enable vector search and then enable streams with KEYS_ONLY (or enable both at CreateTable time), no error is generated - but the vector store doesn’t seem to find anything. Perhaps the vector store can’t pick up the events because the streams code changed the CDC's configuration to not output deltas or post-images. 

The problem is that both the Alternator Streams and Vector Search features make use of CDC, and want to enable CDC in different ways with different parameters.

In this series we add extensive tests that check how the two features work together (they don't, before the fix), and then fix the bug. Fixing it in all the places was more difficult than I anticipated, but the result is very satisfying: The vector search and streams features are now completely independent, each can be enabled or disabled at will, and the user cannot be able to tell that under the hood both use the same ScyllaDB CDC mechanism.

Well, **almost completely** independent. There is one subtle interaction:

After we fixed #7239, you can disable a Stream and it remains visible and readable. But now, if a vector index is **also** enabled, we can't let the user continue to read the old stream - it won't really be an "old stream" any more - it will continue to be updated with new content that the vector store needs, something which shouldn't happen on a disabled stream. So in this specific case, the disabled stream will NOT be visible to the user.

Fixes SCYLLADB-1909
Fixes SCYLLADB-2051

### CODE REVIEW

**Comment by nyh** [comment]
This pull request is incomplete, and therefore opened as a draft. Right now it contains just reproducing tests - not a fix.

**Comment by nyh** [comment]
This PR is now complete (include both tests and the actual fix) so I dropped it out of "draft" mode.

**Comment by nyh** [comment]
Again, the bot's analysis is 100% correct. I'll fix the bug.

File: alternator/executor.cc [hunk 1]
```diff
@@ -1898,16 +1933,54 @@ future<executor::request_return_type> executor::update_table(client_state& clien
...
+                    // we need the CDC options to have enabled=false (CDC is
+                    // explicitly enabled) and delta_mode=full (vector store
+                    // requirement). is_vector_only_cdc() will return true
+                    // with this setting, so DescribeTable/ListStreams will
+                    // not show this CDC as a stream.
+                    if (!tab->indices().empty()) {
+                        cdc::options opts;
+                        opts.enabled(false);
+                        opts.set_delta_mode(cdc::delta_mode::full);
+                        opts.ttl(tab->cdc_options().ttl());
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor.cc:2002]: I think you're right. I'll write a test for this and fix it.

[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor.cc:2002]: Actually, you were wrong. If vector index is enabled, there is no "disabled stream" that we need to preserve - there is an active stream which has it's own setup (specifically, no preimage/postimage) and we can't and don't want to save the old setup.

File: alternator/executor.cc [hunk 2]
```diff
@@ -2048,6 +2120,10 @@ future<executor::request_return_type> executor::update_table(client_state& clien
                     index_options["dimensions"] = std::to_string(dimensions);
                     builder.with_index(index_metadata{index_name, index_options,
                             index_metadata_kind::custom, index_metadata::is_local_index(false)});
+                    // If the table already has CDC-enabled streams using
+                    // delta_mode=keys, upgrade to delta_mode=full so the vector
+                    // store can read full column data from the CDC log.
+                    upgrade_cdc_delta_for_vector_index(builder);
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/executor.cc:2147]: Good catch. I added yet another test check that confirms this bug (it was very hard to notice, because it only affects one very specific scenario, and even then it's just an efficiency issue, not a correctness issue. I'll fix it.

**Comment by ScyllaPiotr** [comment]
> After we fixed #7239, you can disable a Stream and it remains visible and readable. But now, if a vector index is **also** enabled, [...] the disabled stream will NOT be visible to the user.

TBH I don't like this regression. The fact that it is not being documented by this PR is secondary.

**Comment by nyh** [comment]
> > After we fixed #7239, you can disable a Stream and it remains visible and readable. But now, if a vector index is **also** enabled, [...] the disabled stream will NOT be visible to the user.
> 
> TBH I don't like this regression. The fact that it is not being documented by this PR is secondary.

It's not a "regression" in the sense that it does NOT happen in the cases that you considered - without a vector index.
But unfortunately, this is an inherent part of the design that the CDC is **shared** between Streams and Vector Index. We just cannot disable - no matter how much we hope we could - the stream while a vector index still uses it. We can - and do - "hide" the stream from the user, tell the user the stream is disabled, but we cannot let the user continue to read the old stream thinking it was closed. We can't close it.

Another option could have been to forbid the user from disabling the stream. But this would have been even worse, in my opinion. Also, what if the stream had options (like OLD_IMAGE or NEW_IMAGE) that the vector index doesn't even need? Why not allow the user to disable those?

I can think how/where to document this delicate point, but I think it cannot be fixed as long as the approach is that both features (Streams and Vector Search) use CDC under the hood.

**Comment by ScyllaPiotr** [comment]
When Global Tables are enabled for a DynamoDB table, Streams are used for the underneath sync. So Streams are required to be enabled, and the behavior is (and I hope Opus is correct here):
- Streams are visibly enabled — When a table is part of a global table, the Stream details section shows Streams as enabled with NEW_AND_OLD_IMAGES stream view type.
- Users cannot disable Streams — DynamoDB prevents you (both in the console and via API) from turning off Streams or changing the stream settings while the table participates in a global table replica set.
- Streams are auto-enabled on global table creation — When you create a global table, DynamoDB automatically enables Streams with the required settings.

If you want to disable Streams, you must first remove the table from the global table (delete the replica), and only then can you disable Streams.

So I vote to make Streams explicitly (visibly) enabled on the user's side too whenever VS is used. It would cause much less confusion and not limit on the functionality side.

**Comment by nyh** [comment]
Very interesting I didn't know any of what you wrote about global tables. I'll write test to confirm. If DynamoDB already has a feature which "forces" streams, I agree we can use the same approach, although I do believe that my approach is better and more convenient for users and can also be more efficient (not forcing specific delta/images/etc.) that you don't need.

**Comment by ScyllaPiotr** [comment]
> can also be more efficient (not forcing specific delta/images/etc.) that you don't need.

We could be adjusting the Streams type to the current needs.

**Comment by ScyllaPiotr** [comment]
> I'll write test to confirm.

You could also play around inside a browser.

**Comment by nyh** [comment]
> > I'll write test to confirm.
> 
> You could also play around inside a browser.

As the saying goes, "Let a man play around inside a browser, and he'll have answers for today. Teach a man to write a test, he'll ensure Alternator's compatibility with DynamoDB forever" ;-) If eventually, we want to support the Global Tables API, we'll need to understand it anyway.

**Comment by nyh** [comment]
The test that failed in CI is a recently-added Alternator-vector-test test from https://github.com/scylladb/scylladb/pull/29554, the CI ran the entire suite 100 (!) times and it only failed once, and the failure is in pytest_asyncio/plugin.py's attempt to create a connection, so I suspect Scylla crashed in the middle (but I don't know how to find the log...) or perhaps even more likely, this is a test framework flakiness that has nothing to do with this patch.

Regretfully, I'll have to debug this now, we can't just ignore it.

**Comment by nyh** [comment]
> Regretfully, I'll have to debug this now, we can't just ignore it.

Opened https://scylladb.atlassian.net/browse/SCYLLADB-2053.

I couldn't find the root cause. Seems that the test framework thought Scylla didn't boot (which has nothing to do with Alternator or this PR) - I'm guessing this is not a Scylla crash (I didn't find evidence of that) but a bug in the test framework that tried to use the connection (CQL) too early. But it's just a guess and I don't have hard proof.

**Comment by nyh** [comment]
@ScyllaPiotr I'm now at a crossroads.

As you suggested, I took a look at the DynamoDB Global Tables feature (massive test suite - https://github.com/scylladb/scylladb/pull/29928). As you said, the Global Tables feature does force DynamoDB Streams on, in a fairly intrusive and visible way:
1. Turning on Global Tables turns on Streams with the `OLD_AND_NEW_IMAGE` option.
2. The user can see this stream, and use it.
3. When you finally turn off Global Tables (by removing the extra replicas of the table), the stream is **not** removed automatically.
4. If Streams is already on when Global Tables is turned on, it is left with its original type (`KEYS_ONLY` or any other type). It turns out that Global Tables can use any stream type - not just `OLD_AND_NEW_IMAGE` - they just thought this type will be a useful default.
5. The user is not allowed to disable Stream while the Global Table is in effect.

I can implement Alternator's Vector Index in exactly the same way:
1. If you don't have Stream enabled yet, then enabling vector index will enable a user-visible `KEYS_ONLY` stream. But this will be a slightly unusual `KEYS_ONLY` - it will enable also "full delta" mode which isn't normally enabled on KEYS_ONLY stream but the vector store needs. This should be fine.
2. If you do have Stream enabled already before enabling a vector index, it will verify the stream type is support: NEW_IMAGES and OLD_AND_NEW_IMAGES are allowed. The modes `KEYS_ONLY` and `OLD_IMAGES` could be allowed by forcing on the "full delta" mode as explained above, which will be very unusual but could work.
3. The user is not allowed to disable the stream while vector index exists. It's also not possible to change the stream type because this requires disabling the stream first.
4. When the last vector index on a table is removed, the user-visible stream remains on and is not automatically disabled. We should consider disabling the "full data" mode that is nothing but waste for the Stream API that doesn't read it.

I liked my solution allows the user to basically use each of these features independently without considering (usually) the effects of one or the other. Just because we implemented two features on top of the same low-level feature (CQL's CDC) doesn't mean the user needs to know about that.

But this global-tables-like solution isn't that awful either...

**Comment by ScyllaPiotr** [comment]
> @ScyllaPiotr I'm not at a crossroads.
> 
> As you suggested, I took a look at the DynamoDB Global Tables feature (massive test suite - #29928). As you said, the Global Tables feature does force DynamoDB Streams on, in a fairly intrusive and visible way:
> 1. Turning on Global Tables turns on Streams with the `OLD_AND_NEW_IMAGE` option.
> 2. The user can see this stream, and use it.
> 3. When you finally turn off Global Tables (by removing the extra replicas of the table), the stream is **not** removed automatically.
> 4. If Streams is already on when Global Tables is turned on, it is left with its original type (`KEYS_ONLY` or any other type). It turns out that Global Tables can use any stream type - not just `OLD_AND_NEW_IMAGE` - they just thought this type will be a useful default.
> 5. The user is not allowed to disable Stream while the Global Table is in effect.

:+1: 

> I can implement Alternator's Vector Index in exactly the same way:
> 1. If you don't have Stream enabled yet, then enabling vector index will enable a user-visible `KEYS_ONLY` stream. But this will be a slightly unusual `KEYS_ONLY` - it will enable also "full delta" mode which isn't normally enabled on KEYS_ONLY stream but the vector store needs. This should be fine.

Again, following what Dynamo does, I propose to keep it explicit. If VS requires any higher mode, why not enable it explicitly, co the Streams consumer sees them too? 

> 2. If you do have Stream enabled already before enabling a vector index, it will verify the stream type is support: NEW_IMAGES and OLD_AND_NEW_IMAGES are allowed. The modes `KEYS_ONLY` and `OLD_IMAGES` could be allowed by forcing on the "full delta" mode as explained above, which will be very unusual but could work.

I like the former part, plus let's consider automatic explicit step-up to the nearest compatible type:
- KEYS_ONLY -> NEW_IMAGES
- OLD_IMAGES -> OLD_AND_NEW_IMAGES
 
> 3. The user is not allowed to disable the stream while vector index exists. It's also not possible to change the stream type because this requires disabling the stream first.
> 4. When the last vector index on a table is removed, the user-visible stream remains on and is not automatically disabled. We should consider disabling the "full data" mode that is nothing but waste for the Stream API that doesn't read it.
> 
> I liked my solution allows the user to basically use each of these features independently without considering (usually) the effects of one or the other. Just because we implemented two features on top of the same low-level feature (CQL's CDC) doesn't mean the user needs to know about that.
> 
> But this global-tables-like solution isn't that awful either...

Your solution was almost perfect, the only thing I did not like about it was the functionality regression - you essentially reverted my fix for https://github.com/scylladb/scylladb/issues/7239

**Comment by nyh** [comment]
> Your solution was almost perfect, the only thing I did not like about it was the functionality regression - you essentially reverted my fix for #7239

But that's not really true - you still have the ability to use a disabled Stream. You just can't use a disabled Stream in the case that you *also* use a vector index.
Note that in your latest proposal, there's an even bigger limitation: if you're using a stream *and* vector index, you're not allowed to use KEYS_ONLY  or OLD_IMAGES mode. Isn't this an even bigger regression?
Not to mention the regression that you're not allowed to disable Stream if you have a vector index?

**Comment by ScyllaPiotr** [comment]
> > Your solution was almost perfect, the only thing I did not like about it was the functionality regression - you essentially reverted my fix for #7239
> 
> But that's not really true - you still have the ability to use a disabled Stream. You just can't use a disabled Stream in the case that you _also_ use a vector index. 

Exactly the problem.
> Note that in your latest proposal, there's an even bigger limitation: if you're using a stream _and_ vector index, you're not allowed to use KEYS_ONLY or OLD_IMAGES mode. Isn't this an even bigger regression? Not to mention the regression that you're not allowed to disable Stream if you have a vector index?

Indeed, you are limited to other modes, but they can all serve all the expected information, and more.

**Comment by m-szymon** [comment]
@nyh do you plan any more changes here? Or can I review.

**Comment by denesb** [comment]
@nyh please rebase, there are conflicts.

**Comment by nyh** [comment]
> > I can implement Alternator's Vector Index in exactly the same way:
> > 
> > 1. If you don't have Stream enabled yet, then enabling vector index will enable a user-visible `KEYS_ONLY` stream. But this will be a slightly unusual `KEYS_ONLY` - it will enable also "full delta" mode which isn't normally enabled on KEYS_ONLY stream but the vector store needs. This should be fine.
> 
> Again, following what Dynamo does, I propose to keep it explicit. If VS requires any higher mode, why not enable it explicitly, co the Streams consumer sees them too?

Because, the DynamoDB API doesn't have the "delta" mode which our CQL CDC have, and the vector index can use.

Yes, I can enable CDC's "post-image" mode - which does have a DynamoDB API equivalent and that the vector index can also use - but this is **less efficient** than delta mode (it requires every update to list the entire item even if one attribute changed). So your suggestion will make it implementation less efficient. Isn't this a problem?

> > 2. If you do have Stream enabled already before enabling a vector index, it will verify the stream type is support: NEW_IMAGES and OLD_AND_NEW_IMAGES are allowed. The modes `KEYS_ONLY` and `OLD_IMAGES` could be allowed by forcing on the "full delta" mode as explained above, which will be very unusual but could work.
> 
> I like the former part, plus let's consider automatic explicit step-up to the nearest compatible type:
> 
>     * KEYS_ONLY -> NEW_IMAGES
> 
>     * OLD_IMAGES -> OLD_AND_NEW_IMAGES

Yes, I can do that, but we lose on efficiency.
 
> > 3. The user is not allowed to disable the stream while vector index exists. It's also not possible to change the stream type because this requires disabling the stream first.
> > 4. When the last vector index on a table is removed, the user-visible stream remains on and is not automatically disabled. We should consider disabling the "full data" mode that is nothing but waste for the Stream API that doesn't read it.
> > 
> > I liked my solution allows the user to basically use each of these features independently without considering (usually) the effects of one or the other. Just because we implemented two features on top of the same low-level feature (CQL's CDC) doesn't mean the user needs to know about that.
> > But this global-tables-like solution isn't that awful either...
> 
> Your solution was almost perfect, the only thing I did not like about it was the functionality regression - you essentially reverted my fix for #7239

As I commented above, I don't think this is a fair representation of what is really going on with my implementation. The nice thing about my original implementation was that it made it possible to present to the user that the impression that the Vector Search and DynamoDB Streams features are completely independent - you can turn on one or the other or both, then disable one or the other, and everything works correctly. There is just **one** very esoteric (in my opinion) snag: If you disable streams but keep vector search enabled, because we don't really disable CDC, we can't continue to allow you to read the old CDC stream - otherwise you'll see there new things you don't expect to see there.

**Comment by nyh** [comment]
I've converted this PR to a draft. I couldn't convince @ScyllaPiotr that what I did regarding the interaction of vector search and streams was good, so I ended up making various parallels to how vector search "should" behave to how an existing feature on DynamoDB - global tables - behave.

However, now that DynamoDB actually introduced their vector search, this is no longer hypothetical - we can check exactly how their vector search and streams can or can't be enabled together, and our ideal implementation would be to exactly replicate this experience. I'll pick up the tests that I wrote for this PR and move them to the new Alternator Vector Search PR in https://github.com/scylladb/scylladb/pull/31181, and if the tests fail we'll need to fix this in a new PR that should come after https://github.com/scylladb/scylladb/pull/31181 is in, and replace this PR.

**Comment by ScyllaPiotr** [comment]
> I've converted this PR to a draft. 
> [...]
> I'll pick up the tests that I wrote for this PR and move them to the new Alternator Vector Search PR in #31181, and if the tests fail we'll need to fix this in a new PR that should come after #31181 is in, and replace this PR.

So maybe you wanted to close this PR instead.

**Comment by nyh** [comment]
> > I've converted this PR to a draft.
> > [...]
> > I'll pick up the tests that I wrote for this PR and move them to the new Alternator Vector Search PR in #31181, and if the tests fail we'll need to fix this in a new PR that should come after #31181 is in, and replace this PR.
> 
> So maybe you wanted to close this PR instead.

But it still has good code I haven't moved to other PRs. https://github.com/scylladb/scylladb/pull/31181 right now only copied the tests from this PR, but not the code fix.
Soon I'll probably move the code fixes too and then I'll close this PR.

**Comment by nyh** [comment]
> > > I've converted this PR to a draft.
> > > [...]
> > > I'll pick up the tests that I wrote for this PR and move them to the new Alternator Vector Search PR in #31181, and if the tests fail we'll need to fix this in a new PR that should come after #31181 is in, and replace this PR.
> > 
> > 
> > So maybe you wanted to close this PR instead.
> 
> But it still has good code I haven't moved to other PRs. #31181 right now only copied the tests from this PR, but not the code fix. Soon I'll probably move the code fixes too and then I'll close this PR.

With AI, "soon" becomes immediately ;-) I'm closing this PR.

**Comment by nyh** [comment]
> With AI, "soon" becomes immediately ;-) I'm closing this PR.

@ScyllaPiotr to clarify, I know you had a problem with this PR, namely that:

1. If you have a stream and then disable it, you should be able to read the old stream for an extra 24 hours. This old stream is readable and won't continue to get updated.
2. However, with this PR, if you have **both** vector index and stream, and disable the stream, you won't be able to read it any more. If you were allowed to read it, you'd discover that it kept getting updated after you disabled it (and confusingly, with different postimage/preimage settings).

I will mention this new limitation in the new PR, and will add an xfailing test demonstrating it, but I honestly believe that the ability to **use** both features together without limitation is far more important than some corner case of what happens after you disable one of the features. I still want it documented (xfailing test), but I don't want to block progress because of it.

**Comment by ScyllaPiotr** [comment]
> > With AI, "soon" becomes immediately ;-) I'm closing this PR.
> 
> @ScyllaPiotr to clarify, I know you had a problem with this PR, namely that:
> 1. If you have a stream and then disable it, you should be able to read the old stream for an extra 24 hours. This old stream is readable and won't continue to get updated.
> 2. However, with this PR, if you have **both** vector index and stream, and disable the stream, you won't be able to read it any more. If you were allowed to read it, you'd discover that it kept getting updated after you disabled it (and confusingly, with different postimage/preimage settings).
> 
> I will mention this new limitation in the new PR, and will add an xfailing test demonstrating it, but I honestly believe that the ability to **use** both features together without limitation is far more important than some corner case of what happens after you disable one of the features. I still want it documented (xfailing test), but I don't want to block progress because of it.

I'm yet to see which approach you chose (and this depends on how Dynamo works in detail), but what I advocated was to be explicit - if a user asked to disable Streams but some "mechanism A" still needs the Streams active to in order to function, don't close the Streams - keep them open, only rearrange them to the minimum info they need to gather.
